### PR TITLE
Fix converting only float type tensors in Lite

### LINKED ISSUE #10371: Why does bf16 of LightningLite use GPU Memory much more than pytorch_lightning.Trainer
## 🐛 Bug

I test two ways to train my model, LightningLite and pytorch_lightning.Trainer on 2 A100 machines with the same configuration, The Memory used is very different.
- LightningLite(bf16): about 24GB
- LightningLite(fp16): about 27GB
- LightningLite(fp32): about 27GB
- LightningLite(deepspeed, fp16): about 14GB
- pytorch_lightning.Trainer(bf16): about 17GB

### To Reproduce
```python
class Lite(LightningLite):
    def run(self, model, optimizer, lr_scheduler, logger, num_files=20, global_step=0):
        self.setup(model, optimizer)
        while True:
            fps = self.get_fps(model.cfg)
            for i in range(len(fps) // num_files):
                train_loader = self.get_data(model.cfg, i, fps, num_files)
                train_loader = self.setup_dataloaders(train_loader)
                model.train()
                metric = {'loss': 0.0, 'mlm_acc': 0.0, 'sop_acc': 0.0, 'log_step': 0.0}
                start_time = time.time()
                for j, batch in enumerate(tqdm(train_loader)):
                    outputs = model(**batch, output_hidden_states=True)
                    labels = batch['labels']

                    loss = outputs.loss
                    optimizer.zero_grad()
                    self.backward(loss)
                    optimizer.step()
                    lr_scheduler.step()
                    global_step += 1

                    if global_step >= model.cfg.SOLVER.MAX_ITERS:
                        break
                if global_step >= model.cfg.SOLVER.MAX_ITERS:
                    break
            if global_step >= model.cfg.SOLVER.MAX_ITERS:
                break

    def get_fps(self, cfg):
        """
        get files path for train.
        """
        if ',' in cfg.DATASETS.TRAIN:
            dirs = cfg.DATASETS.TRAIN.split(',')
            dirs = [d.strip() for d in dirs]
        else:
            dirs = [cfg.DATASETS.TRAIN]
        fps = []
        for d in dirs:
            _fns = os.listdir(get_abs_path(d))
            _fns = [fn for fn in _fns if fn[:4] in ['part', 'ppar']]
            _fps = [get_abs_path(d, fn) for fn in _fns]
            fps += _fps
        random.shuffle(fps)
        return fps

    def get_data(self, cfg, idx, fps, num=20):
        parts = get_parts(fps, idx, num)
        train_loader, _, _ = make_loaders(cfg, fps=parts, num_proc=8)
        return train_loader


def get_parts(fps, i, n):
    if i > len(fps) // n:
        i = 0
    elif i == len(fps) // n and len(fps) % n == 0:
        i = 0
    return fps[i * n:(i + 1) * n]


def main():
    cfg = args_parse('deberta/deberta_v2.yml')
    model = BertsForPretraining(cfg)
    optimizer = make_optimizer(cfg, model)
    lr_scheduler = build_lr_scheduler(cfg, optimizer, offset=0)['scheduler']
    logger = getLogger(cfg.MODEL.NAME)
    logger.setLevel(logging.DEBUG)
    gpus = cfg.MODEL.GPUS if cfg.MODEL.GPUS != 0 else cfg.MODEL.GPU_IDS
    lite = Lite(
        strategy='ddp',
        gpus=gpus,
        accelerator='gpu',
        precision='bf16',
        num_nodes=2
    )
    try:
        lite.run(model, optimizer, lr_scheduler, logger, 20, 2650000)
    except:
        logger.error(traceback.format_exc())


if __name__ == '__main__':
    seed_everything()
    main()
```

### Environment

- PyTorch Lightning Version (e.g., 1.3.0): 1.5.0
- PyTorch Version (e.g., 1.8): 1.10+cu113
- Python version: 3.8.12
- OS (e.g., Linux): ubuntu20.04
- CUDA/cuDNN version: 11.4
- Device: 2 machines, 8*A100 each.
- GPU models and configuration: 
- How you installed PyTorch (`conda`, `pip`, source):
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### Additional context
Here is env dockerfile.

```Dockerfile
FROM nvcr.io/nvidia/pytorch:21.09-py3
RUN apt-get update && apt-get -y install openssh-server htop tmux \
    && mkdir -p /var/run/sshd
RUN pip install -U pytorch-lightning transformers yacs ujson pkuseg pypinyin deepspeed datasets tqdm wandb && pip uninstall -y torch && pip install torch==1.10.0+cu113 torchvision==0.11.1+cu113 torchaudio==0.10.0+cu113 -f https://download.pytorch.org/whl/cu113/torch_stable.html && pip uninstall -y torchtext
```

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #10371 

Lite converts inputs automatically to the target precision (32, 16, bf16). However, so far this happened unconditionally for all tensor types, when in reality this only makes sense for floating point tensors. This PR restricts this conversion to float tensors only, both for input and output of the wrapped model-forward.

Test fails on master (i.e., without the fix applied).

### Does your PR introduce any breaking changes? If yes, please list them.

No

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

Part of #1 (it's a lie, this is just here to avoid noisy GitHub bot)

### CODE REVIEW

**Review by kaushikb11 — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/lite/wrappers.py [hunk 1]
```diff
@@ -95,12 +95,17 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
         }
         # TODO (@awaelchli): let the precision plugin handle the conversion
         to_type = precision_to_type[precision]
-        args, kwargs = apply_to_collection([args, kwargs], function=lambda t: t.to(to_type), dtype=Tensor)
+
+        def _convert_float_tensor(t: Tensor) -> Tensor:
+            return t.to(to_type) if torch.is_floating_point(t) else t
```
Comment (tchaton) [inline pytorch_lightning/lite/wrappers.py:100]: noob question: If you provide both float32, and int32 and request precision=16 or "bf16": torch.bfloat16. Would this convert only the float32 -> float16 but leave the int32 unchanged ?

Comment (carmocca) [reply pytorch_lightning/lite/wrappers.py:100]: Yes. This is why #10079 is important, an example:

```python
import torch

torch.set_default_dtype(torch.double)
x = torch.tensor(1)
print(x.dtype)  # int64
```

which is what you would expect.

Comment (carmocca) [reply pytorch_lightning/lite/wrappers.py:100]: Also, this should probably differentiate between `AMP` 16, bf16 vs full 16, bf16 conversion

**Review by carmocca — APPROVED** [review]
We'll need to iterate on this but it's okay for now

File: pytorch_lightning/lite/wrappers.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/lite/wrappers.py:100]: That's great, but #10079 and Carlos's explanation is more about the creation of tensors if I understand correctly. Here we are concerned about the conversion going in and out of forward. Where the tensors are created is out of control of Lite (could be inside or outside run). If #10079 lands it can be done for Lite easily.
